### DISCUSSION TITLE
Eraser Brush Runtime Error

### DISCUSSION THREAD

**Discussion by dilankamr** [fabricjs/fabric.js#8339] [General]
Hi, 

I am using a fabric.js custom build by changing some codes. I have included **EraserBrush** to my build by importing **./src/mixins/eraser_brush.mixin** to my **index.js** file. 

Here is the built fabric.js fie: [https://raw.githubusercontent.com/dilankamr/fabric.js/master/dist/fabric.js](https://raw.githubusercontent.com/dilankamr/fabric.js/master/dist/fabric.js)

And, I have used EraserBrush in my application as below. (as mentioned in fabric.js demos: [http://fabricjs.com/erasing](http://fabricjs.com/erasing))

 ```
            this.canvas.freeDrawingBrush = new fabric.EraserBrush(this.canvas);
            this.canvas.isDrawingMode = true; 
```

But, when I going to use the eraser, **I am receiving the below errors while clicking on the canvas**. 

ERROR TypeError: Cannot read properties of undefined (reading '_isMainEvent')_ at klass.onMouseDown
ERROR TypeError: this.freeDrawingBrush.onMouseMove is not a function at klass._onMouseMoveInDrawingMode

<img width="666" alt="image" src="https://user-images.githubusercontent.com/10356798/193451853-9b6b016b-88a7-42cd-8457-26c5751be295.png">

I debugged the issue by adding some breakpoints, 

I found **this keyword hasn't required attributes** and therefore **this.canvas gets undefined** when executing **onMouseDown** function. 

<table>
<tr>
<td>
<img width="513" alt="image" src="https://user-images.githubusercontent.com/10356798/193454086-55b8e051-618a-4e6e-a750-f4e8ee1e2e3a.png">
</td>
<td>
<img width="513" alt="image" src="https://user-images.githubusercontent.com/10356798/193453604-8ef7a540-a278-490f-92b2-a04beeb8f42a.png">
</td>
</tr>
</table>

And, 

**this.freeDrawingBrush hasn't required attributes** and therefore **this.freeDrawingBrush.onMouseMove gets undefined** when executing **_onMouseMoveInDrawingMode** function.

<table>
<tr>
<td>
<img width="513" alt="image" src="https://user-images.githubusercontent.com/10356798/193454136-2c698f4b-5785-46a0-ae62-0fd6b4641245.png">
</td>
<td>
<img width="513" alt="image" src="https://user-images.githubusercontent.com/10356798/193454148-8871ad4e-0917-4bb5-a77c-409010198ab6.png">
</td>
<td>
<img width="513" alt="image" src="https://user-images.githubusercontent.com/10356798/193453622-93c80e11-95bb-40db-ae4f-d71e9467ea20.png">
</td>
</tr>
</table>

Could you please help me to resolve this issue?

Thanks,
Dilanka Rathnayake.

**Comment by ShaMan123** [comment]
answered #8262
